File: frontend/src/types/order.ts
export interface OrderItem {
  productId: string;
  quantity: number;
  price: number;
}

export interface OrderItemRequest {
  productId: string;
  quantity: number;
  price: number;
}

export interface OrderItemResponse {
  id: number;
  productId: string;
  quantity: number;
  price: number;
  name: string;
}

export interface ShippingAddress {
  shippingAddress: string;
  city: string;
  district: string;
  ward: string;
  postalCode?: string;
  phoneNumber: string;
}

export interface OrderRequest extends ShippingAddress {
  userId: string;
  orderItems: OrderItemRequest[];
  couponCode?: string;
}

export interface OrderResponse extends ShippingAddress {
  id: string;
  userId: string;
  items: OrderItemResponse[];
  status: 'PENDING' | 'PROCESSING' | 'SHIPPED' | 'DELIVERED' | 'CANCELLED';
  createdAt: string;
  lastModifiedDate: string;
  subtotal: number;
  discountAmount: number;
  couponCode?: string;
  totalPrice: number;
  shippingFee: number;
}

export interface OrderPreviewResponse {
  userId: string;
  items: OrderItemResponse[];
  subtotal: number;
  discountAmount: number;
  couponCode?: string;
  shippingFee: number;
  totalPrice: number;
}

export interface Coupon {
  code: string;
  discountPercent: number;
  active: boolean;
  minimumOrderAmount: number;
  expiryDate?: string;
  createdAt: string;
  updatedAt: string;
}

File: backend/src/main/java/com/shopcart/order/entity/Order.java
package com.shopcart.order.entity;

import com.shopcart.common.enums.OrderStatus;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "orders")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
public class Order {
    @Id
    @Column(name = "order_id", length = 50)
    private String id; 

    @Column(name = "user_id", nullable = false)
    private String userId;

    @Column(name = "total_price", nullable = false)
    private Long totalPrice; 

    @Column(name = "shipping_fee")
    private Long shippingFee;

    @Column(name = "discount_amount")
    @Builder.Default
    private Long discountAmount = 0L;  // Discount in VND

    @Column(name = "coupon_code", length = 50)
    private String couponCode;

    @Column(name = "shipping_address", length = 255)
    private String shippingAddress;

    @Column(name = "city", length = 100)
    private String city;

    @Column(name = "district", length = 100)
    private String district;

    @Column(name = "ward", length = 100)
    private String ward;

    @Column(name = "postal_code", length = 20)
    private String postalCode;

    @Column(name = "phone_number", length = 20)
    private String phoneNumber;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status; 

    @Column(name = "created_at", updatable = false)
    @CreatedDate
    private LocalDateTime createdAt;

    @Column(name = "last_modified_date")
    @LastModifiedDate
    private LocalDateTime lastModifiedDate;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<OrderItem> orderItems = new ArrayList<>();
}